src/foam/nanos/pm/FoldReducePMLogger.java

```diff
@@ -6,116 +6,154 @@
 
 package foam.nanos.pm;
 
-import foam.core.ContextAwareSupport;
+import foam.core.FObject;
 import foam.core.X;
 import foam.dao.DAO;
-import foam.dao.MDAO;
 import foam.dao.ProxyDAO;
+import foam.dao.RemoveSink;
 import foam.mlang.order.Comparator;
 import foam.mlang.predicate.Predicate;
 import foam.util.concurrent.FoldReducer;
-import java.util.HashMap;
-import java.util.Map;
+
+import java.util.*;
 
 public class FoldReducePMLogger
   extends    FoldReducer
   implements PMLogger
 {
+  private final Set<PMInfo> captureEnabledList_ = new TreeSet<PMInfo>(PMInfo.ID);
 
   public FoldReducePMLogger() {
   }
 
-
   /** Template method to Create initial state. **/
   public Object initialState() {
-    return new HashMap();
-  }
+    HashMap<String, PMInfo> initState = new HashMap<>();
+
+    // Load capture enabled PMInfos into initial state
+    if ( captureEnabledList_ != null && captureEnabledList_.size() > 0 ) {
+      for ( PMInfo pmi_orig : captureEnabledList_ ) {
+        PMInfo pmi = new PMInfo();
+        pmi.setKey(pmi_orig.getKey());
+        pmi.setName(pmi_orig.getName());
+        pmi.setCapture(true);
+
+        String key = pmi_orig.getKey() + ":" + pmi_orig.getName();
+        initState.put(key, pmi);
+      }
+    }
 
+    return initState;
+  }
 
   /** Template method to Fold a new update into a state. **/
   public void fold(Object state, Object op) {
-    Map map = (Map) state;
+    Map<String, PMInfo> map = (Map<String, PMInfo>) state;
     PM  pm  = (PM) op;
 
     // Regular PMInfo
-    Object key = pm.getKey() + ":" + pm.getName();
-    PMInfo pi  = (PMInfo) map.get(key);
-
-    if ( pi == null ) {
-      pi = new PMInfo();
-      pi.setKey(pm.getKey());
-      pi.setName(pm.getName());
-      map.put(key, pi);
+    String key = pm.getKey() + ":" + pm.getName();
+    PMInfo pmi  = map.get(key);
+
+    if ( pmi == null ) {
+      pmi = new PMInfo();
+      pmi.setKey(pm.getKey());
+      pmi.setName(pm.getName());
+      map.put(key, pmi);
     }
 
-    pi.fold(pm);
-    if ( pi.getCapture() ) capture(pi);
+    pmi.fold(pm);
+    if ( pmi.getCapture() ) {
+      capture(pmi);
+    }
   }
 
+  public void capture(PMInfo pmi) {
+    StringBuilder trace = new StringBuilder("// Captured at: ");
+    trace.append(new Date());
+    trace.append(System.lineSeparator());
 
-  public void capture(PMInfo pi) {
-    StringBuilder trace = new StringBuilder();
     for ( StackTraceElement j : Thread.currentThread().getStackTrace() ) {
       trace.append(j.toString());
-      trace.append(System.getProperty("line.separator"));
+      trace.append(System.lineSeparator());
     }
 
-    pi.setCapture(false);
-    pi.setCaptureTrace(trace.toString());
+    pmi.setCapture(false);
+    pmi.setCaptureTrace(trace.toString());
+    updateCaptureEnabledList(pmi);
```

Comment:
On line 84 would should cause the FoldReducer to reduce. This will prevent all of the other threads from still trying to capture as well.

Reply:
Added getState() to force reduce

Reply:
it ended up being line 86 but hopefully you wont hold that against me